apis/duck/v1alpha1/sinkable_types.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Sinkable is very similar concept as Targetable. However, at the
+// transport level they have different contracts and hence Sinkable
+// and Targetable are two distinct resources.
+// It would be better to have one level of indirection from Status.
+// The way this is currently put in at the same level as the other Status
+// resources. Hence the objects supporting Sinkable (Knative channel for
+// example), will expose it as: Status.DomainInternal
+// For new resources that are built from scratch, they should probably
+// introduce an extra level of indirection.
+package v1alpha1
+
+import (
+	"encoding/json"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"github.com/knative/pkg/apis/duck"
+)
+
+// Sinkable is the schema for the sinkable portion of the payload
+type Sinkable struct {
+	DomainInternal string `json:"domainInternal,omitempty"`
+}
+
+// Implementations can verify that they implement Sinkable via:
+var _ = duck.VerifyType(&Sink{}, &Sinkable{})
+
+// Sinkable is an Implementable "duck type".
+var _ duck.Implementable = (*Sinkable)(nil)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Sink is a skeleton type wrapping Sinkable in the manner we expect
+// resource writers defining compatible resources to embed it.  We will
+// typically use this type to deserialize Sinkable ObjectReferences and
+// access the Sinkable data.  This is not a real resource.
+type Sink struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Status Sinkable `json:"status"`
```

Comment:
I understand that you have Targetable as Status for legacy reasons, but you also indicate:
> // For new resources that are built from scratch, they should probably
> // introduce an extra level of indirection.

Maybe Sinkable also suffers the legacy burden of wanting to support Route/Service?  (see comment above about how this is different from Targetable)

Given that Eventing hasn't yet released, compatibility with knative/serving 0.1 is (likely?) not a requirement and we could modify knative/serving at HEAD to move the targetable stuff you need around.  We couldn't drop `status.domainInternal` (yet), but we could add a status field for `targetable`, `sinkable`, as needed.

I think my preference would be to be more prescriptive about `.status.targetable` and (quickly) move `knative/serving` at HEAD to implement that instead of living with legacy baggage forever.  WDYT?

Reply:
Yeah, I send you that observation last night offline as well. I also made a comment to that effect on Sinkable about the extra level of indirection.
I think we should do it. However, I'd propose to change that in a subsequent PR so that I could make the change in Serving first, then backport this code to support that?